draft-ietf-privacypass-integration.md

```diff
@@ -0,0 +1,147 @@
+ipr: trust200902
+keyword: Internet-Draft
+
+stand_alone: yes
+pi: [toc, sortrefs, symrefs]
+
+author:
+ -
+    ins: A. Davidson
+    name: Alex Davidson
+    org: Brave Software
+    city: Lisbon
+    country: Portugal
+    email: alex.davidson92@gmail.com
+
+normative:
+  RFC2119:
+  RFC8446:
+  I-D.irtf-cfrg-voprf:
+  I-D.ietf-privacypass-architecture:
+  I-D.ietf-privacypass-http-api:
+informative:
+  RFC7049:
+  RFC7159:
+
+--- abstract
+
+This document specifies the necessary integration for building the
+Privacy Pass protocol, supporting symmetric verification and public
+metadata, on top of an existing oblivious pseudorandom function
+protocol.
+
+--- middle
+
+# Introduction
+
+The Privacy Pass protocol provides a privacy-preserving authorization
+mechanism. In essence, the protocol allows clients to provide
+cryptographic tokens that prove nothing other than that they have been
+authorized by a given server in the past
+{{I-D.ietf-privacypass-architecture}}.
+
+This document provides the necessary integration for building the
+authorization framework, based on existing constructions of oblivious
+pseudorandom function protocols {{I-D.irtf-cfrg-voprf}}. Moreover, we
+show how this integration allows `public` metadata to be introduced to
+the protocol, that is agreed by both clients and servers.
+
+This document DOES NOT cover the architectural framework required for
+running and maintaining the Privacy Pass protocol in the Internet
+setting. In addition, it DOES NOT cover the choices that are necessary
+for ensuring that client privacy leaks do not occur. Both of these
+considerations are covered in {{I-D.ietf-privacypass-architecture}}. In
+addition, considerations of how to embed the protocol interactions in
+the HTTP setting are considered in {{I-D.ietf-privacypass-http-api}}.
+
+# Terminology
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in {{RFC2119}}.
+
+The following terms are used throughout this document.
+
+- Client: An entity that provides authorization tokens to services
+  across the Internet, in return for authorization.
+- Server: A service (also known as an Issuer) that issues Privacy Pass
+  tokens to clients.
+- Key: The secret key used by the server for authorizing client data.
+- Blinding/unblinding: A pair of cryptographic operations that allow
+  randomly distributing and revealing some underlying `data`, whilst
+  preserving the ability to generate valid signatures on `data` itself.
+
+We assume that all protocol messages are encoded into raw byte format
+before being sent across the wire.
+
+# Privacy Pass flow {#protocol-flow}
+
+There are three phases in the protocol: the initialization phase, the
+issuance phase, and the redemption phase.
+
+In the initialization phase, the server samples a keypair and publishes
+a ``public key`` (or key commitment) to a public repository.
```

Comment:
```suggestion
a key configuration in a way that clients can retrieve it. This configuration consists
of the server's public key and configuration information for the underlying POPRF.

~~~
struct {
   uint8 version;
   uint16 suite;
   uint8 public_key[Ne];
} KeyConfig;
~~~

KeyConfig.suite corresponds to a POPRF ciphersuite from {{I-D.irtf-cfrg-voprf, Section 4}},
and KeyConfig.public_key corresponds to a serialized public key of length `Ne` bytes 
(denoted as a `SerializedElement` in {{I-D.irtf-cfrg-voprf, Section 2}}).
```

Reply:
It's hard to interop if we don't describe how configurations are represented, I think.

Reply:
We should probably also indicate how the server knows which KeyConfig to use for a given request. I don't think this needs to be part of the protocol here, as it may vary. Instead, I'd recommend coming up with a unique way to identify a KeyConfig, e.g., id = SHA256(KeyConfig), and then saying that the wrapper protocol around Privacy Pass can send this to the server if needed for the given application.

Reply:
Version of this key config, so it should probably also include a length of the struct? 🤷